ndb_api/clone.go

```diff
@@ -15,136 +15,53 @@ package ndb_api
 
 import (
 	"context"
-	"encoding/json"
 	"errors"
 	"fmt"
-	"io"
 	"net/http"
 
 	"github.com/nutanix-cloud-native/ndb-operator/ndb_client"
 	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 // Fetches all the clones on the NDB instance and retutns a slice of the databases
-func GetAllClones(ctx context.Context, ndbClient *ndb_client.NDBClient) (clones []DatabaseResponse, err error) {
+func GetAllClones(ctx context.Context, ndbClient ndb_client.NDBClientHTTPInterface) (clones []DatabaseResponse, err error) {
 	log := ctrllog.FromContext(ctx)
-	log.Info("Entered ndb_api.GetAllClones")
-	if ndbClient == nil {
-		err = errors.New("nil reference: received nil reference for ndbClient")
-		log.Error(err, "Received nil ndbClient reference")
+	if _, err = sendRequest(ctx, ndbClient, http.MethodGet, "clones?detailed=true", nil, &clones); err != nil {
+		log.Error(err, "Error in GetAllClones")
 		return
 	}
-	res, err := ndbClient.Get("clones?detailed=true")
-	if err != nil || res == nil || res.StatusCode != http.StatusOK {
-		if err == nil {
-			if res != nil {
-				err = fmt.Errorf("GET /clones responded with %d", res.StatusCode)
-			} else {
-				err = fmt.Errorf("GET /clones responded with a nil response")
-			}
-		}
-		log.Error(err, "Error occurred fetching all clones")
-		return
-	}
-	log.Info("GET /clones", "HTTP status code", res.StatusCode)
-	body, err := io.ReadAll(res.Body)
-	defer res.Body.Close()
-	if err != nil {
-		log.Error(err, "Error occurred reading response.Body in GetAllClones")
-		return
-	}
-	err = json.Unmarshal(body, &clones)
-	if err != nil {
-		log.Error(err, "Error occurred trying to unmarshal.")
-		return
-	}
-	log.Info("Returning from ndb_api.GetAllClones")
 	return
```

Comment:
All the functions in ndb_api package were performing same nil checks for ndb client, empty params, http calls, data deserialization and error handling. Now, all the common logic has moved to a `sendRequest` function.